Opt-In Policy For More Robust Fee-bumping

glozow opened the pull request:
A specification is useful for coordination between node impls that want to implement the same policy and applications that want to use it. For those that are not interested in the details of v3 policy, this also serves as a writeup of the specific pinning problems we aim to address. There has been discussion of using this in other protocol design and multiple requests for its documentation to exist in the BIPs repository, so I'm opening a PR here.

Implementation:
- https://github.com/bitcoin/bitcoin/pull/28948
- https://github.com/bitcoin/bitcoin/pull/28984

Example usage and things built on top:
- Ephemeral Anchors
  - https://github.com/bitcoin/bitcoin/pull/29001
  - https://github.com/bitcoin/bips/pull/1524
- LN commitment transactions
  - https://delvingbitcoin.org/t/lightning-transactions-with-v3-and-ephemeral-anchors/418
  - https://github.com/instagibbs/bolts/commits/zero_fee_commitment
- LN-Symmetry
  - https://delvingbitcoin.org/t/ln-symmetry-project-recap/359
  - https://github.com/instagibbs/lightning/tree/eltoo_support

Discussion and history:
- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-January/019817.html
- https://gist.github.com/glozow/25d9662c52453bd08b4b4b1d3783b9ff
- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-September/020937.html
- https://delvingbitcoin.org/t/v3-transaction-policy-for-anti-pinning/340
- https://github.com/bitcoin/bitcoin/pull/25038

t-bast commented:
FWIW, concept ACK :)

murchandamus commented on bip-v3.mediawiki:
The last sentence makes use of Rule 4 which is only defined below. Perhaps the order of these two rules should be swapped?

glozow replied:
Consolidated the rules a bit to be more understandable and not have these referencing problems.

murchandamus reviewed (commented):
ACK 4bd12d5698404e28c70d4486316f4e1cca356896

glozow commented:
Thanks @murchandamus, took all your suggestions.

ismaelsadeeq commented on bip-v3.mediawiki:
The implementation is now using the fee rate diagram for incentive compatibility check

glozow replied:
Made the language more ambiguous

ismaelsadeeq reviewed (commented):
Concept ACK

glozow commented:
@luke-jr this has been open for a month, would you mind taking a look?

ariard commented:
I think this should document that 1000 vb child limit is experimental and it cannot be relied on by downstream projects.
This opt-in policy is not robust towards NTA pinning (cf. “The Good, The Bad, The Ugly” 2020 mail) and “loophole” pinning exposed in Core’s #28948. This can note that additional opt-in policy might be applied on top of nversion=3 see Core’s #29454.

murchandamus reviewed (commented):
ACK 63e8a7191f8ab2453e5971e62f709df613ae224c

glozow commented on bip-truc.mediawiki:
taken

glozow commented on bip-truc.mediawiki:
taken

glozow commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,271 @@
+<ref>
+[https://github.com/bitcoin/bitcoin/blob/632a2bb731804dffe52bd4cbd90bfee352d25ede/doc/policy/mempool-replacements.md Bitcoin Core's RBF policy] at the time of writing. It is slightly different from what is described in BIP 125.
+</ref>
+to keep the more incentive compatible transaction, i.e. one that would earn a miner more fees. As such, creating higher feerate double-spends (replacements) is often employed by users as a fee-bumping mechanism.
```
taken

glozow commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,271 @@
+to limit the computational complexity of mempool operations and mitigate Denial of Service attacks.
+
+Users may also create unconfirmed transactions that conflict with -- or are "double spends" of -- each other by spending the same input(s) in both.
+Instead of always keeping the first transaction, many mempools also have some kind of Replace by Fee (RBF) policy
```
taken

glozow commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,271 @@
+
+==Abstract==
+
+Users can set <code>nVersion=3</code> on a transaction to opt in to restrictions on spending unconfirmed outputs (making them Topologically Restricted Until Confirmation or TRUC transactions) in exchange for improved fee-bumping reliability. Nodes apply a different set of mempool policies to these transactions. Namely, topological restrictions that make it easier to assess the incentive compatibi ...
```
took some of this

glozow reviewed (commented):
Thanks @murchandamus! Accepted your suggestions.

murchandamus commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,271 @@
+  Status: Draft
+  Type: ???
+  Created: 2024-01-10
+  License: BSD-3-Clause
```
Please add the "Post-History" header in the preamble and link to relevant mailing list threads and other discussions (e.g. delving).

glozow replied:
done, thanks

luke-jr commented:
As with #1524, node policy is not a standardizable subject matter in itself - every node decides its own policy. BIP 125 already defined a way for wallets to indicate they prefer RBF-like treatment of their transactions, and I don't see why the same signal can't be used for this.

vostrnad commented:
Every node deciding its own policy is not mutually exclusive with it being standardizable – each node decides which (if any) standards to follow. Policy standards make just as much sense as e.g. wallet standards or P2P protocol standards.

Roasbeef commented:
> BIP 125 already defined a way for wallets to indicate they prefer RBF-like treatment of their transactions, and I don't see why the same signal can't be used for this.

BIP 125 defines a set of rules w.r.t transaction replacement. Some of those rules have edge cases that can [lead to defects where transactions that should otherwise be replaced, can't be replaced](https://github.com/bitcoin/bips/blob/158acdbbbf8ef13f6b345b6281a96e88e20d2cf9/bip-truc.mediawiki#user-content-RBF_pinning_through_Rule_3). This class of replacement defects loosely falls under the umbrella of "transaction pinning". The TRUC rules resolve those issues, creating a replacement semantics that better serve off-chain protocols, and address some known pinning vectors. For many off-chain protocols, TRUC will supersede base RBF. As the semantics are distinct, it cannot use the existing sequence fields carved out by BIP 125 (the rules are also incompatible). Instead, it uses transaction v3 (no longer non standard) as a way to signal replacement under a distinct set of rules.

luke-jr commented:
BIP 125 defines a signalling mechanism, not policy (which is again outside the scope of BIPs) even if it describes a particular policy. The intent of the signal is clear, and it can be implemented in other ways such as the one that seems to be desired here.

Roasbeef commented:
> BIP 125 defines a signalling mechanism, not policy

It clearly defines _both_: 

> The opt-in full Replace-by-Fee (opt-in full-RBF) signaling **policy** described here allows spenders to add a signal to a transaction indicating that they want to be able to replace that transaction in the future


> This **policy** specifies two ways a transaction can signal that it is replaceable.


> Inherited signaling: Transactions that don't explicitly signal replaceability are replaceable **under this policy** for as long as any one of their ancestors signals replaceability and remains unconfirmed.


> Because descendant transactions may also be replaceable **under this policy** through inherited signaling, any method used to process opt-in full-RBF transactions should be inherited by any descendant transactions for as long as any ancestor opt-in full-RBF transactions remain unconfirmed.

It also defines [a precise set of rules for the policy](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#implementation-details). If an implementation deviates from those rules, then it isn't BIP 125. From the OP, there's a clear need to define a new policy _and_ signalling mechanism, which this document does.

luke-jr commented:
It may be poorly worded, but that isn't the point.

luke-jr commented:
(Note that BIP 125's relationship with policy *was* a matter of discussion back when it was submitted, and it was nearly rejected - the only reason it was accepted in the end, was as a signalling BIP. And even if you want to insist it defines policy, mistakes made in the past would not be a reason to repeat them in the future, and thus would actually work against accepting this proposal)

glozow commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,271 @@
+<pre>
+  BIP: ???
+  Layer: Applications
+  Title: Topologically Restricted Until Confirmation Transactions
```
changed. 44 is not a lot of characters!

glozow commented:
Addressed @murchandamus's comments and slightly restructured the text.

> As with https://github.com/bitcoin/bips/pull/1524, node policy is not a standardizable subject matter in itself - every node decides its own policy.

I really don't see how the text in this BIP can be interpreted to mean that nodes can't decide their own policy. But I've added some language to the text to make it more clear that it's optional, that there are other ways to do this, policy is not consensus, and the details can be different.

> BIP 125 already defined a way for wallets to indicate they prefer RBF-like treatment of their transactions, and I don't see why the same signal can't be used for this.

This policy restriction should be opt in to avoid breaking existing use cases. Also I don't agree that the same signal should be used for 2 different things just because they have a few things in common.

ariard commented:
> (Note that BIP 125's relationship with policy was a matter of discussion back when it was submitted, and it was nearly
> rejected - the only reason it was accepted in the end, was as a signalling BIP. And even if you want to insist it defines
> policy, mistakes made in the past would not be a reason to repeat them in the future, and thus would actually work against > accepting this proposal)

If I can make one suggestion it would be to split this BIP in two components:
- the signaling mechanism (nversion=3)
- the set of policy rules (i.e the TRUC)

This would certainly help to navigate more efficiently this kind of situation in the future.

"Similarly, there are multiple approaches to creating a policy to minimize pinning, more may become available over time (see Related Work section), and the details of this approach can be tweaked if conditions change. For example, if loosening one of the topology restrictions enables a new use case while still providing acceptable pinning bounds, it can be changed.”

This gives flexibility to have many signaling mechanism (e.g nversion / nsequence) committing to one set of policy rules and vice-versa have one signaling mechanism committing to many policy rules (bip125, 1000 vb child limit truc, etc). This can only give us more flexibility w.r.t coordinated upgrades across bitcoin layers.

murchandamus commented on bip-truc.mediawiki:
```suggestion
Senders can signal that they want a transaction to be Topologically Restricted Until Confirmation (TRUC). Specifically, set <code>nVersion=3</code>.
```

murchandamus reviewed (approved):
My concerns have been addressed, I just noticed a typo you could fix in case you need to touch this again.

reACK f2d33b5cf01e04e190a3493a74c220fb64266ee8

glozow commented:
Thanks for the review @jonatack! I'll wait for a number assignment.

Dragankamil6 commented on bip-truc.mediawiki:
```diff
@@ -0,0 +1,286 @@
+  Author: Gloria Zhao <gloriajzhao@gmail.com>
+  Comments-URI:
+  Status: Draft
+  Type: ???
```
bip-truc.mediawiki

Dragankamil6 reviewed (approved):
Dragankamil6@gmail.com

ProofOfKeags commented on bip-truc.mediawiki:
Is "sigop-adjusted virtual size" formally defined anywhere. I'm aware of the sigop limit and virtual size. I am unaware of how sigops may or may not "adjust" the virtual size.

glozow replied:
Not really, apart from the code comments written in/since its introduction in https://github.com/bitcoin/bitcoin/pull/8365. This is one of the main reasons why I think policy ought to be documented more formally, as I and others have tripped on this quite a bit.

Also see https://github.com/bitcoin/bitcoin/pull/27591 (which I still need to get around to updating - up for grabs if you're interested).

ProofOfKeags replied:
Got it. Reading that link indicates that on a *policy* level Core attaches a vByte price to sigops when evaluating mempool/relay stuff. However, at the *consensus* level we just need to have a sigops limit as well as a minimum vByte threshold per sigop?

glozow replied:
Correct. It's simplifying a 2D Knapsack problem by rolling both values into 1 score.

I'm going to mark this thread as resolved as I don't think this  BIP would be the right place for defining it. Please feel free to comment on the above PR, open a stack exchange question, or DM me if you have further questions.

glozow commented:
Thanks @jonatack @murchandamus!

I've added the number, and used that to update the file name, readme table, etc. Also took all of your suggestions.

glozow commented on bip-0431.mediawiki:
```diff
@@ -0,0 +1,291 @@
+  Layer: Applications
+  Title: Topology Restrictions for Pinning
+  Author: Gloria Zhao <gloriajzhao@gmail.com>
+  Comments-URI:
```
That fixed it, thanks!

carlaKC reviewed (approved):
ACK 32e4e44

With the tiny non-blocking nitpick that a few places in the doc use markdown `_` for italics (I think) where they should use `'' ` for [mediawiki italics](https://www.mediawiki.org/wiki/Help:Formatting#:~:text=You%20can%20format%20your%20text,apostrophes%20like%20''this''%20).

murchandamus commented on bip-0431.mediawiki:
The syntax for unordered lists in the mediawiki format uses asterisks:

```suggestion
* https://github.com/bitcoin/bitcoin/pull/28948
* https://github.com/bitcoin/bitcoin/pull/29873
* https://github.com/bitcoin/bitcoin/pull/29496
```

glozow replied:
done

murchandamus commented on bip-0431.mediawiki:
```suggestion
<br/> Consider a rule where the fee can be decreased (remove Rule 3 and 4) but the feerate must double. In this scenario, a 100 kvB transaction can be replaced by a 100 vB transaction paying 200 sats. That's 200 sats to relay 100,200 vB of transaction data, which is less than 0.002 sat/vB. It becomes quite cheap to replace large portions of the mempool, decreasing both its average feerate and total absolute fees.
```

murchandamus commented on bip-0431.mediawiki:
```suggestion
<br/>Consider a rule where the fee can stay the same (keep Rule 3 but drop Rule 4) but the feerate must double. The attacker can start out with 100 kvB transaction, paying 1 sat/vB. A user can reduce its size over and over again, doubling the feerate each time until it gets too small, and end up paying 100 ksat for 100 kvB (1 + 1/2 + 1/4 + ... + log2(mintxsize)) -> approaches 200 kvB. This means the attacker pays a rate of 0.5 sat/vB to relay transactions, which is below our "free relay" threshold of 1 sat/vB.
```

murchandamus commented on bip-0431.mediawiki:
```suggestion
<br />This rule is also easily tacked on to existing logic for policy and wallets. A maximum size standard transaction (100 kvB) can have up to 1000 vB of descendants to be within the default descendant limit (101 kvB).
```

murchandamus commented on bip-0431.mediawiki:
```suggestion
<br/>The smaller this limit, the fewer UTXOs a child may use to fund this fee-bump. For example, only allowing the TRUC child to have 2 inputs would require wallets to maintain a pool of high-value confirmed UTXOs. However, as the fee-bumping child only needs to fund fees (as opposed to payments), just a few UTXOs should suffice. With a limit of 1000 vB and usage of taproot outputs, the child can have 15 inputs and 2 outputs (calculated using [https://bitcoinops.org/en/tools/calc-size/ this tool]).
```

murchandamus reviewed (approved):
This looks good to me and ready to go except for a couple nits (which I don’t think would need to hold up a merge) and a formatting issue:

1. The JEDEC standard specifies that uppercase _K_ (kilo) refers to 1024¹ (and _M_ (mega), _G_ (giga) as 1024² and 1024³ respectively). However, since the numbers in your example are referring to a factor of 1000¹, they should be using lowercase _k_, the symbol of the [metric prefix kilo](https://en.wikipedia.org/wiki/Kilo-) referring to a factor of 1000¹. Therefore, it would be preferable if the units were updated to read "kvB" instead of "KvB", and "ksats" instead of "Ksats".

2. AFAIU, technical writing standards generally recommend a space between the numerical value and unit symbol.

E.g. using a space is recommended by the SI Standard,
![image](https://github.com/bitcoin/bips/assets/4060799/68498d41-6b3a-4b7b-941e-62544d40dcb7) <sup>via https://www.nist.gov/pml/special-publication-330/sp-330-section-5</sup>

by NIST,
![image](https://github.com/bitcoin/bips/assets/4060799/9ec4913d-ef6f-4432-9ed1-bae719067573) <sup>via https://physics.nist.gov/cuu/Units/checklist.html</sup>

as well as IEEE:
![image](https://github.com/bitcoin/bips/assets/4060799/5c63b24d-27ed-45d5-a0c8-695ed977b3db) <sup>via https://academia.stackexchange.com/q/54885/8305</sup>

My personal preference is to separate the value and symbol with a [Narrow No-Break Space (U+202F)](https://www.compart.com/en/unicode/U+202F) as that provides the expected visual offset, renders as a smaller gap than a full length space, but still prevents unfortunate line breaks. My suggested changes already make use of NNBSPs.

murchandamus reviewed (commented):
This looks good to me and ready to go except for a formatting issue and a couple nits (which I don’t think would need to hold up a merge):

1. The JEDEC standard specifies that uppercase _K_ (kilo) refers to 1024¹ (and _M_ (mega), _G_ (giga) as 1024² and 1024³ respectively). However, since the numbers in your example are referring to a factor of 1000¹, they should be using lowercase _k_, the symbol of the [metric prefix kilo](https://en.wikipedia.org/wiki/Kilo-) referring to a factor of 1000¹. Therefore, it would be preferable if the units were updated to read "kvB" instead of "KvB", and "ksats" instead of "Ksats".

2. AFAIU, technical writing standards generally recommend a space between the numerical value and unit symbol.

E.g. using a space is recommended by the SI Standard,
![image](https://github.com/bitcoin/bips/assets/4060799/68498d41-6b3a-4b7b-941e-62544d40dcb7) <sup>via https://www.nist.gov/pml/special-publication-330/sp-330-section-5</sup>

by NIST,
![image](https://github.com/bitcoin/bips/assets/4060799/9ec4913d-ef6f-4432-9ed1-bae719067573) <sup>via https://physics.nist.gov/cuu/Units/checklist.html</sup>

as well as IEEE:
![image](https://github.com/bitcoin/bips/assets/4060799/5c63b24d-27ed-45d5-a0c8-695ed977b3db) <sup>via https://academia.stackexchange.com/q/54885/8305</sup>

My personal preference is to separate the value and symbol with a [Narrow No-Break Space (U+202F)](https://www.compart.com/en/unicode/U+202F) as that provides the expected visual offset, renders as a smaller gap than a full length space, but still prevents unfortunate line breaks. My suggested changes already make use of NNBSPs.

glozow commented on bip-0431.mediawiki:
done

glozow commented on bip-0431.mediawiki:
done

glozow commented on bip-0431.mediawiki:
done

sdaftuar commented on bip-0431.mediawiki:
The reference to "Rule 3" (and "Rule 5" and "Rule 6" below) might be confusing to readers who haven't read BIP 125 -- perhaps add a link or explanation?

glozow replied:
I've renamed the headings to be more descriptive, and added links and clarifications where "Rule n" is mentioned throughout the document

sdaftuar replied:
Looks better to me, thanks!

glozow commented:
(I think this is ready, please lmk if there's anything left to do on my end)

petertodd commented:
This BIP is missing an explanation of how it is expected to work with future nVersion upgrades.

Eg if we add a third consensus critical transaction version, how should that interact with TRUC? Do we just add another version?

Obviously this is a serious drawback of mixing up mempool behavior and consensus code.

murchandamus commented:
I’m not sure I understand your concern. If someone e.g. proposed giving meaning to version 4 in say a _BIP-v4_, it could propose how transactions with version 3 should interact with transactions of version 4 and then whoever implements BIP-v4 would follow those recommendations, while those that don’t implement it, would not.

petertodd commented:
TRUC behavior is not desirable for all transactions; it is not a clear upgrade.

Thus, the obvious question is how do we expect a future soft fork version upgrade to deal with this?

For example, consider what would have happened had we tried to introduce TRUC prior to v2 transactions.

This can of course be avoided by not using a version number to signal TRUC behavior. I do not see a reason why we actually need to do that, given how narrow the actual usage of TRUC is likely to be. Triggering it simply on zero fee transactions should be fine.

On July 8, 2024 4:58:22 PM GMT+02:00, murchandamus ***@***.***> wrote:
>I’m not sure I understand your concern. If someone e.g. proposed giving meaning to version 4 in say a _BIP-v4_, it could propose how transactions with version 3 should interact with transactions of version 4 and then whoever implements BIP-v4 would follow those recommendations, while those that don’t implement it, would not.
>

murchandamus commented:
> TRUC behavior is not desirable for all transactions; it is not a clear upgrade. Thus, the obvious question is how do we expect a future soft fork version upgrade to deal with this? For example, consider what would have happened had we tried to introduce TRUC prior to v2 transactions.

It’s not clear to me where you see a problem. This BIP applies to transactions with version 3, not transactions signaling a version of 3 and higher.

petertodd commented:
So an obvious example of where TRUC's use of v3 transactions would cause a problem is if we had a situation similar to v2 transactions, where an upgrade happened that had wide applicability. Since TRUC is so limited, we'd need to have two different versions of the consensus upgrade for TRUC and non-TRUC transactions. Obviously, that would be a mess. This gets even worse when you consider that we all know that TRUC only mitigates pinning for a very narrow use-case, and people are already discussing extensions to TRUC with new version bits.

Meanwhile, it would be quite easy for TRUC to achieve it's actual main goal of enabling empheral anchor outputs as quick fix by defining TRUC behavior as being enabled for zero-fee transactions.

Choosing V3 didn't even do "TRUC behavior" cleanly as a bit: bin(3) = 0b11, two bits set.

ariard commented:
> Choosing V3 didn't even do "TRUC behavior" cleanly as a bit: bin(3) = 0b11, two bits set.

Let’s say we have a hypothetical soft-fork in the future to clean all the pinning and transaction-relay jamming broken mess, nVersion is a 32-bit field, so we can still split it in two 16-bit halves: one set for truc-like policy behavior and one set for consensus validation.

Without an example, it’s quite theoretical as there are always other signaling options: the taproot annex, the input nsequence, the nlocktime field...

ProofOfKeags commented:
> one set for truc-like policy behavior and one set for consensus validation.

I do think that bifurcating things along consensus vs policy is likely a good idea. It would fully orthogonalize the problems. While consensus-invalid implies policy-invalid, I believe these are otherwise independent concerns and we should treat them as such in their signaling/encoding schemes.

glozow commented:
My understanding was that the version field *is* the place for policy/application signals *because* it's not expected to be reinterpreted in the future for consensus things. That is why I chose it.
